tests/Doctrine/Tests/ORM/PersistentCollectionTest.php

```diff
@@ -144,6 +144,16 @@ public function testClearWillAlsoResetKeyPositions()
         $this->assertEquals([0], array_keys($this->collection->toArray()));
     }
 
+    public function testClearWillMakeCollectionDirty()
```

Comment:
@syastrebov can you add a test like in https://github.com/doctrine/doctrine2/tree/master/tests/Doctrine/Tests/ORM/Functional/Ticket to show the higher level impact of this patch? I'm specifically wondering about how this affects `flush()` semantics.

An example as a response to this comment is also sufficient to evaluate this.

Reply:
@Ocramius added the functional test, please take a look if it's what you expected

Reply:
Thanks for that! I agree with the patch here, but it changes behaviour in a quite aggressive way. @Majkl578 does this count as a BC break in your opinion?

Reply:
I'm a bit concerned about possible impact, especially if this changes what or how events are fired during `flush()`.
Also there is no test added to cover the changes to these events although the PR aims to _"fix pre update lifecycle callbacks for MANY_TO_MANY collections"_.